plugins/superio/fu-superio-device.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+#include <string.h>
+#include <sys/errno.h>
+
+#include <glib/gstdio.h>
+
+#include "fu-superio-device.h"
+
+#define FU_PLUGIN_SUPERIO_TIMEOUT	5 /* s */
+
+#define KB_IBF			(1 << 1)  /* i/p buffer full */
+#define KB_OBF			(1 << 0)  /* o/p buffer full */
+
+struct _FuSuperioDevice {
+	FuDevice		 parent_instance;
+	gchar			*devport;
+	gint			 fd;
+	guint16			 data_port;
+	guint16			 cmd_port;
+	guint16			 id;
+	guint32			 size;
+};
+
+G_DEFINE_TYPE (FuSuperioDevice, fu_superio_device, FU_TYPE_DEVICE)
+
+static void
+fu_superio_device_to_string (FuDevice *device, GString *str)
+{
+	FuSuperioDevice *self = FU_SUPERIO_DEVICE (device);
+	g_string_append (str, "  FuSuperioDevice:\n");
+	g_string_append_printf (str, "    fd:\t\t\t%i\n", self->fd);
+	g_string_append_printf (str, "    devport:\t\t%s\n", self->devport);
+	g_string_append_printf (str, "    data-port:\t\t0x%04x\n", (guint) self->data_port);
+	g_string_append_printf (str, "    cmd-port:\t\t0x%04x\n", (guint) self->cmd_port);
+	g_string_append_printf (str, "    id:\t\t\t0x%04x\n", (guint) self->id);
+	g_string_append_printf (str, "    size:\t\t0x%04x\n", (guint) self->size);
+}
+
+gboolean
+fu_superio_device_open (FuSuperioDevice *self, GError **error)
+{
+	/* open device */
+	self->fd = g_open (self->devport, O_RDWR);
+	if (self->fd < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to open %s: %s",
+			     self->devport,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_outb (FuSuperioDevice *self, guint16 port, guint8 data, GError **error)
+{
+	if (pwrite (self->fd, &data, 1, (goffset) port) != 1) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to write to port %04x: %s",
+			     (guint) port,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_inb (FuSuperioDevice *self, guint16 port, guint8 *data, GError **error)
+{
+	if (pread (self->fd, data, 1, (goffset) port) != 1) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to read from port %04x: %s",
+			     (guint) port,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static guint16
+fu_superio_device_probe_id (FuSuperioDevice *self, GError **error)
+{
+	guint8 msb;
+	guint8 lsb;
+
+	if (!fu_superio_device_outb (self, 0x2e, 0x20, error))
+		return FALSE;
+	if (!fu_superio_device_inb (self, 0x2f, &msb, error))
+		return FALSE;
+	if (!fu_superio_device_outb (self, 0x2e, 0x21, error))
+		return FALSE;
+	if (!fu_superio_device_inb (self, 0x2f, &lsb, error))
+		return FALSE;
```

Comment:
Are these standard commands and addresses that all EC's should respond to?

Reply:
Yes, this is common to all super-io hardware.